### PR TITLE
Add API tests

### PR DESCRIPTION
**Pull request by sotojuan**
This helps out with #285 by simply taking some of the CLI tests and applying them to the API. Some were omitted because they tested output.

### CODE REVIEW

File: test/api.js [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+		path.join(__dirname, 'fixture/one-pass-one-fail.js')
+	];
+
+	var api = new Api(files);
+
+	api.run()
+		.then(function () {
+			t.same(api.passCount, 2);
+			t.same(api.failCount, 1);
+			t.match(api.errors[0].title, /test › fixture › one-pass-one-fail › this is a failing test/);
```
Comment (jamestalmage) [inline test/api.js]: This assertion is failing on AppVeyor (reliably). Are you able to test on Windows?
Unfortunately, the tap `spec` reporter does not show the input string for failed `t.match` tests.
I've been burned by this before.

Maybe it's time we switched to the `tap` `classic` reporter. It's the default reporter on desktop (the default when no `tty` is present is TAP. It is basically `dots`, but with really verbose failures)

**Comment by sotojuan** [comment]
I can test on Windows during work tomorrow. I'll let you know!

**Comment by sotojuan** [comment]
@jamestalmage Just tested on Windows. The error is trivial. I am testing for:

```
test › fixture › one-pass-one-fail › this is a failing test
```

But on Windows it's:

```
test » fixture » one-pass-one-fail » this is a failing test
```

(Notice the error!).

Should be easy to fix.

File: test/api.js [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+		path.join(__dirname, 'fixture/one-pass-one-fail.js')
+	];
+
+	var api = new Api(files);
+
+	api.run()
+		.then(function () {
+			t.same(api.passCount, 2);
+			t.same(api.failCount, 1);
+			t.match(api.errors[0].title, /test (›|») fixture (›|») one-pass-one-fail (›|») this is a failing test/);
```
Comment (vdemedes) [inline test/api.js]: Instead of `(›|»)`, you should use: `figures.pointerSmall` (and require `figures` module at the top).

**Comment by vdemedes** [comment]
@sotojuan Great, I like this PR! Thank you for your contribution!

File: test/api.js [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+	var api = new Api([path.join(__dirname, 'fixture/fail-fast.js')], {
+		failFast: true
+	});
+
+	api.run()
+		.then(function () {
+			t.ok(api.failFast);
+			t.same(api.passCount, 1);
+			t.same(api.failCount, 1);
+			t.true(/false fail false/.test(api.errors[0].error.message));
```
Comment (jamestalmage) [inline test/api.js]: Not related to this PR, but I find `false fail false` to be a very awkward message.

**Comment by jamestalmage** [comment]
:+1: from me.

File: test/api.js (hunk 2 above)
Comment (sindresorhus) [inline test/api.js]: Or just use a loose regex.

File: test/api.js (hunk 3 above)
Comment (sindresorhus) [inline test/api.js]: I noticed this the other day too. Agree. Can you open an issue about improving it?

**Comment by sindresorhus** [comment]
:+1: Looks good :)

Shouldn't we remove the CLI tests that are now tested in the API instead?

**Comment by sotojuan** [comment]
Thanks a lot for the help. I'm not very experienced with regex—what would be a good loose regex to use?

**Comment by sindresorhus** [comment]
@sotojuan Doesn't really matter what the separator is, so I would go with:

```js
/test \w fixture \w one-pass-one-fail \w this is a failing test/
```

**Comment by sotojuan** [comment]
`\w` wasn't working but `\S` (not whitespace) did.

**Comment by vdemedes** [comment]
@sindresorhus fair point. @sotojuan could you please remove the old CLI tests you replaced with your API tests?

**Comment by vdemedes** [comment]
@sotojuan and the `git rebase master` is also needed to merge this ;)

**Comment by sotojuan** [comment]
Sure. Not sure why it's failing though ~_~.

**Comment by jamestalmage** [comment]
> Sure. Not sure why it's failing though ~_~.

The CI builds are failing because you need to rebase.

**Comment by sotojuan** [comment]
Hm not very experienced with rebase, sorry. What should I be doing?

**Comment by sindresorhus** [comment]
https://www.google.no/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=fix%20merge%20conflict

**Comment by sindresorhus** [comment]
There's a conflict because of https://github.com/sindresorhus/ava/commit/386416846bdb7b8aae84e1b0d8cf77da0bb31fa4.

**Comment by sotojuan** [comment]
All right, figured it was time to learn more git stuff. Fixed it (I think), will push again soon.

File: test/fixture/source-map-pragma-exception.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+const test = require('../../');
```
Comment (sindresorhus) [inline test/fixture/source-map-pragma-exception.js]: I don't think you meant to include this file. It was already removed in https://github.com/sindresorhus/ava/commit/386416846bdb7b8aae84e1b0d8cf77da0bb31fa4.

File: test/api.js [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+		t.true(/Can\'t catch me!/.test(data.message));
+	});
+
+	api.run()
+		.then(function () {
+			t.same(api.passCount, 1);
+		});
+});
+
+test('stack traces for exceptions are corrected using a source map', function (t) {
```
Comment (sindresorhus) [inline test/api.js]: > stack traces for exceptions are corrected using a source map **file**

File: test/api.js [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+		t.match(data.stack, /^.*?at\b.*source-map-initial-input.js:7.*$/m);
+	});
+
+	api.run()
+		.then(function () {
+			t.same(api.passCount, 1);
+		});
+});
+
+test('stack traces for exceptions are corrected using a source map, found via a pragma', function (t) {
```
Comment (sindresorhus) [inline test/api.js]: This test was removed in https://github.com/sindresorhus/ava/commit/386416846bdb7b8aae84e1b0d8cf77da0bb31fa4

File: test/api.js [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+var Api = require('../api');
+
+test('ES2015 support', function (t) {
+	t.plan(1);
+
+	var api = new Api([path.join(__dirname, 'fixture/es2015.js')]);
+
+	api.run()
+		.then(function () {
+			t.same(api.passCount, 1);
```
Comment (sindresorhus) [inline test/api.js]: This should use `t.is`, `t.same` is for deep testing.

**Comment by vdemedes** [comment]
LGTM

**Comment by vdemedes** [comment]
@sotojuan new changes got merged into master branch, so this PR needs rebase one more time.

**Comment by sindresorhus** [comment]
LGTM

**Comment by sindresorhus** [comment]
@vdemedes No need. There were no conflicts.

**Comment by sindresorhus** [comment]
@sotojuan Thanks again for this awesome contribution! 👌 So nice not having all those CLI tests anymore :) Keep the great pull requests coming.

tumblr_m4rz1h2lw21qdlh1io1_400

**Comment by vdemedes** [comment]
@sindresorhus Oh, github's message then was not unclear. I wonder if it's related to lgtm.co.

**Comment by sindresorhus** [comment]
@vdemedes Yes, it was missing one LGTM.

**Comment by vdemedes** [comment]
@sotojuan Thank you for your work!

**Comment by sotojuan** [comment]
No problem, and thank you guys for your help and patience!

**Pull request closed without merging**